lib/ansible/parsing/vault/__init__.py

```diff
@@ -878,10 +879,8 @@ def _edit_file_helper(self, filename, secret,
         # An existing vaultfile will always be UTF-8,
         # so decode to unicode here
         b_ciphertext = self.vault.encrypt(b_tmpdata, secret, vault_id=vault_id)
-        self.write_data(b_ciphertext, tmp_path)
+        self.write_data(b_ciphertext, filename)
```

Comment:
This creates a concurrency issue, now the file is unsafe to use as someone might be editing it while someone else is trying to use it in a play.

To do this change safely you must introduce locking to both the editing and the reading of vault files.

Reply:
Thank you for pointing out, indeed there is a chance of a read/write that can conflict with concurrent and incomplete operation.

I would like to try adding a thin locking layer. It will be based on the following observation: all vault operations first read its contents into the memory using `VaultEditor.read_data()`, then process it in the memory, then write results back using `VaultEditor.write_data()`. The edit operation does the same cycle twice: decrypting into temporary plaintext file, then encrypting back into the original vault file.

Thus locking will need to be introduced between `read_data()` and `write_data()` methods. It will not try to lock the vault file itself due to platform compatibility difficulties. It will operate via additional lock file in vim style (e.g. `.vault_filename.lock`). It will make sure that lock file is deleted on error via `atexit` handler. An unsuccessful locking attempt will wait for 5 seconds and then fail globally due to a timeout.

Please tell if you see any problems with this approach. Please also tell if the branch should be rebased on top of the current head.

Reply:
I have many issues with lockfiles, for one `atexit` cannot guarantee their disappearance, only on clean shutdown of the program.

You also need to lock when ansible/ansible-console/ansible-playbook read the files for runtime use, or they can get a incomplete data stream, which is why we use rename, which is atomic and ensures either old or new are read and no 'in between' state is accessible.

Reply:
You express two concerns: 1) that particular locking approach is difficult to implement properly and 2) that, even being implemented in `ansible-vault`, locking still cannot ensure data consistency.

The second concern seems to be more important, so let us model the worst case. `ansible` / `ansible-console` / `ansible-playbook` do not write to the vault, so the fail scenario seems to be the following:

* vault functionality has been designed to support keeping playbooks in version control systems,
* with VCS the normal cycle is: change the vault file, commit the playbook, release, checkout new playbook version onto the controller host, execute,
* despite this design the user changes vault file directly on the controller host, while running the playbook that uses it,
* the vault file is large enough for giving significant chance for concurrent reads,
* `ansible-playbook` reads incomplete data and fails.

Considering, that the current implementation raises an error when run by non-root user (recorded in the issue), I think that this kind of failure can be neglected.

If you'd agree then we can discuss more reliable design of locking.

If you have already evaluated such trade-offs and have concluded that the renaming approach is the best possible then please tell. I understand your work load and don't insist on arguing.

Reply:
I would not dismiss any scenario, too many times have i encountered something like you describe (as well as the complete opposite) for every point.

I'm fine that we add locking (i would make it an option, not the default), I just want it to be added properly as otherwise you end up  with a false sense of security and a race condition waiting to happen. Your setup might not break, but we need to consider much more than that.

I did evaluate trade-offs, which is why we generally use rename instead of locks. If using an SCM to keep the file, permission changes would be apparent in the commit, also i would expect the deployment to ensure correct permissions on the target. 

But, as above, that is not always the case, which is why I'm open to adding the feature, to support other setups.

Reply:
Thank you, I understand it as a blessing to try adding locking. It should be free of races, should work on NFS mounts, should not hang on stale lock files and does not need to support running `ansible-vault` on Windows. Is that correct?

And could you please explain a bit better idea of "add locking, making it an option, not the default"? 

The only purpose of locking was to avoid the loss of ownership, which can only happen when user does not own the vault file. Do you mean that `ansible-vault` should support both renaming and locking schemes and only fall back to locking when that is appropriate?

Reply:
or via `--exclusive` switch.

Reply:
The requested change has been implemented in e6585de08e3d8ba49a4fd1fbed76d555d828a3d2 + 9da37a0c0c0765628dee6a4d3c06bf6dcf38a1e2.